src/alloc.c

```diff
@@ -438,12 +437,7 @@ static jl_sym_t *symtab_lookup(jl_sym_t *volatile *ptree, const char *str,
             ptree = &node->left;
         else
             ptree = &node->right;
-        node = *ptree;
-        // We may need a data dependency barrier here to pair with the
-        // cpu write barrier in _jl_symbol so that we won't read a invalid
-        // value from `node`. However, this shouldn't be a problem on all
-        // the architectures we currently support.
-        // https://www.kernel.org/doc/Documentation/memory-barriers.txt
+        node = jl_atomic_load_acquire(ptree);
```

Comment:
Right, it' won't help x86 hardware since regular loads already have acquire semantics.  The weak semantics are not actually implemented because the C++11 definition made it unworkable.  [This paper](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0190r0.pdf) describes the problems in its current definition and how the committee proposes to fix it.